documentation/pattern-match-avoid-duplication.md

```diff
@@ -0,0 +1,266 @@
+    Is42
+    IsNot42(actual_value: u32)
+    IsNot42AndNoValue
+}
+
+fn get_answer() -> Answer { ... }
+
+fn main() {
+    match get_answer() {
+        Is42 => {
+            // long body 1
+        }
+        IsNot42 | IsNot42AndNoValue => {
+            // long body 2
+        }
+    }
+}
+```
+
+In both cases, we want Jakt to ideally only generate the long bodies once.
+
+## A partial improvement (#1597)
+
+PR #1597 addressed Constraint 2.1 by keeping the shared relationship between
+the patterns inside a pattern group and the body they jump to. This enables
+codegen to find pattern groups that have no bindings on them and easily group
+them by concatenating `case` labels or concatenating boolean expressions with
+`||`.
+
+Now each `CheckedMatchCase` contains a list of `CheckedMatchPattern` and one
+body that uses the bindings declared within said pattern. The implementation
+typechecks all bodies like the last version, but only selects the last checked
+body.
+
+The problem with this approach is that now the typechecker portion of this
+solution does not respect Constraint 1, since the last checked body is not
+enough to represent the case if bindings from the same parsed pattern group can
+have different binding types.
+
+This makes `tests/codegen/match_enum_same_body_different_impl.jakt` to fail
+codegen, since it calls `maybe_templated<JaktArray<i32>>` on all cases, when
+the first two cases of the enum should use the overloaded versions of
+`maybe_templated`.
+
+## A better solution (#1602)
+
+With this, Constraint 2.2 is not addressed yet, but it satisfies Constraint 1
+and Constraint 2.1. Typechecker now keeps the grouping, but re-generates the
+bodies when the types don't match.
+
+Each `CheckedMatchCase` contains a list of `CheckedMatchPattern`, a set of
+bindings `[String:VarId]` and a body where those bindings are used. All the
+patterns in this `CheckedMatchCase` are known to have the same set of bindings
+as the case.
+
+### How it works
+
+Typechecker walks the patterns to get the type information out of them as it
+did before, but instead of adding the binding variables to a scope, it adds
+them to a `BindingKeyBuilder`. Typechecker "submits" bindings to the builder,
+which then produces a `BindingKey` that tells us whether the set of bindings
+that were declared already exist within the same pattern group.
+
+Since we don't want to be passing a lot of state regarding the match to helper
+functions like `.typecheck_match_variant`, `BindingKeyBuilder` only contains a
+reference to the slice of cases that have been declared and refer to the same
+`ParsedMatchBody`. Remember that it could be multiple due to Constraint 1.
+
+To enforce that the set of cases checked by `BindingKeyBuilder` are the correct
+set, and that context is not lost when reading through the long loop bodies of
+`.typecheck_match`, all the state regarding match typecheck is in
+`MatchBuilder`. This struct has two functionalities: - It will make us
```

Comment:
where second functionality

Reply:
oh, I incorrectly removed the bullet point for the second paragraph.